src/orc.cpp

```diff
@@ -342,64 +358,162 @@ odrv_report::odrv_report(std::string_view symbol, const die* list_head)
+        // Report everything except the stuff on the ignore list (denylist)
+        return !sorted_has(settings._violation_ignore, category);
+    } else if (!settings._violation_report.empty()) {
+        // Report nothing except the the stuff on the report list (allowlist)
+        return sorted_has(settings._violation_report, category);
+    }
+
+    return true;
 }
 
 /**************************************************************************************************/
-bool filter_report(const odrv_report& report) {
-    std::string odrv_category = report.category();
+// Returns a comma-separated serialization of categories NOT considered when determining an ODRV.
+std::string odrv_report::filtered_categories() const {
+    std::string result;
+    bool first = true;
+
+    for (std::size_t i = 0; i < category_count(); ++i) {
+        auto c = category(i);
+        if (should_report_category(c)) continue;
+
+        if (first) {
+            first = false;
+        } else {
+            result += ", ";
+        }
 
-    // Decide if we should report or ignore.
+        result += std::move(c);
+    }
 
-    auto& settings = settings::instance();
-    bool do_report{true};
+    return result;
+}
 
-    if (!settings._violation_ignore.empty()) {
-        // Report everything except the stuff on the ignore list
-        do_report = !sorted_has(settings._violation_ignore, odrv_category);
-    } else if (!settings._violation_report.empty()) {
-        // Report nothing except the the stuff on the report list
-        do_report = sorted_has(settings._violation_report, odrv_category);
+/**************************************************************************************************/
+// Returns a comma-separated serialization of categories considered when determining an ODRV.
+std::string odrv_report::reporting_categories() const {
+    std::string result;
+    bool first = true;
+
+    for (std::size_t i = 0; i < category_count(); ++i) {
+        auto c = category(i);
+        if (!should_report_category(c)) continue;
+
+        if (first) {
+            first = false;
+        } else {
+            result += ", ";
+        }
+
+        result += std::move(c);
+    }
+
+    return result;
+}
+
+/**************************************************************************************************/
+// Generates a category "slug" based on this symbol's kind + the category (e.g. member:type).
+std::string odrv_report::category(std::size_t n) const {
+    return to_string(_conflict_map.begin()->second._tag) + std::string(":") +
+           (_conflicting_attributes.empty() ? "<none>" : to_string(_conflicting_attributes[n]));
+}
+
+/**************************************************************************************************/
+// Here we have an ODRV report, but need to decide if the ODRV's categories are ones we should
+// report for, or if we should filter this ODRV out of the report.
+bool filter_report(const odrv_report& report) {
```

Comment:
Aren't the two loops superfluous? Do you need to store to "categories", or could you just check as you go?

Reply:
I'm not quite sure which two loops you are talking about. It's very possible one or both could be eliminated.

Reply:
I think he's saying, in the two loops below, instead of gathering all the categories in a vector, and then looping through the vector, you could just move the `should_report_category` check and `return` to the first loop.

Reply:
Fixed - thanks.